### DISCUSSION TITLE
Proxy implementation

### DISCUSSION THREAD

**Discussion by ska-ibees** [eliasdabbas/advertools#199] [Ideas]
Is proxy support currently present?

Well if not:

To avoid being blocked, how about adding something like this [https://github.com/rejoiceinhope/scrapy-proxy-pool](https://github.com/rejoiceinhope/scrapy-proxy-pool)

**Comment by eliasdabbas** [comment]
Great suggestion. This will eventually need to be addressed. 

Would you like to dig into the details and suggest how this can be merged into `advertools.crawl`?

**Reply by ska-ibees** [reply]
Sure thing!

**Comment by ska-ibees** [comment]
Well, after spending plenty of time digging into proxy implementation. 

Below are the findings:

1. General Free proxies are not recommended becuase of performance and security issues.
However, if someone wants to implement ip rotation then better to use high availability IP pools. I have tested https://github.com/SpiderClub/haipproxy/ and it serves the purpose so far.

2. Proxy implementation in advertools
 As of now, I have tested proxy within advertools by hard coding it. 

 ```
def start_requests(self):
        meta = {"proxy": "http://127.0.0.1:3128"}
        for url in self.start_urls:
            try:
                yield Request(url, callback=self.parse, errback=self.errback,meta=meta)
            except Exception as e:
                self.logger.error(repr(e))
```
[Link to file ](https://github.com/ibees-consulting/advertools/blob/75be88ba21532183838684e4ccc5c96d018637ee/advertools/spider.py#L673)

I have been struggling to pass meta into adv.crawl . However, couldn't get it working without monkey-patching base spider class. 
Can you please suggest any way to include proxy within  adv.crawl?

**Reply by eliasdabbas** [reply]
Great stuff!

Did it work as expected when you hard coded it?

If so, the next step is to add the parameter `proxy` to the `crawl` function I guess. Some thoughts I'm still considering and I'd like to know what you think: 

- What would this value be? A single proxy, or would it be a list of proxies?
- How would these proxies be handled, randomly assigned to a call to the crawl function, while following links, etc? 
- How much control would the user need over those proxies? (if this domain, use this proxy, otherwise use that?)


The `crawl` function basically creates a list, which is the sent to the command line using the std lib `subprocess` module. 

For example: 
```python
adv.crawl('https://example.com', 'output.jl')
```

Gets converted to: 

```shell
scrapy runspider /path/to/advertools/spider.py -a url_list=https://example.com -o output.jl
```
This command controls the behavior of the spider, which uses the given arguments to start the crawl process.
It's not very complicated, but not trivial either. 

The main question is how the behavior should work.

@dsottimano any thoughts?

**Reply by ska-ibees** [reply]
> Did it work as expected when you hard coded it?

Yes, i could see request hitting the proxy server. Since i used free IP pool so it wasn't reliable.


> If so, the next step is to add the parameter `proxy` to the `crawl` function I guess. Some thoughts I'm still considering and I'd like to know what you think:
> 
> * What would this value be? A single proxy, or would it be a list of proxies?

Well, there are multiple implementations available (API vs REST vs Manual IP lists etc.) and it depends which type of proxy service is in use. 

> * How would these proxies be handled, randomly assigned to a call to the crawl function, while following links, etc?
Scrapy has built-in support for proxies and generally speaking there are two methods to implement them.

1 . Using Request Parameters
2. Creating a custom Middleware

1 . USING REQUEST PARAMETERS

Generally you just pass an URL and target a callback function when you are using a Scrappy. But, if you are looking at using a specific proxy for a particular Url, then it is possible if you ass a meta tag. Consider this example:

def start_requests(self):

for url in self.start_urls:

return Request(url=url, callback=self.parse,

headers={"User-Agent": "My UserAgent"},

meta={"proxy": "http://192.168.1.1:8050"})

There is a middleware in the Scrapy called Proxy Middleware which passes the request object and sets it up. 

2. CREATE A CUSTOM MIDDLEWARE

In case you do not want to use the middleware that is already provided, the next option that you have is to create a custom middleware. The process of passing this middleware is similar to the one above:

from w3lib.http import basic_auth_header

class CustomProxyMiddleware(object):

def process_request(self, request, spider):

request.meta[“proxy”] = "http://192.168.1.1:8050"

request.headers[“Proxy-Authorization”] =

basic_auth_header(“<proxy_user>”, “<proxy_pass>”)

In the above code we pass the necessary authentication

DOWNLOADER_MIDDLEWARES = {

'myproject.middlewares.CustomProxyMiddleware': 350,

'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware': 400,

}

BTW, I just discovered https://scrapoxy.io/ on this page ([Link](https://docs.scrapy.org/en/latest/topics/practices.html)) . It looks promising. Let me try it out and see how it works... (Update: project is not active :))

> * How much control would the user need over those proxies? (if this domain, use this proxy, otherwise use that?)

I am not sure if Scrapy offers such flexibility. However, Running multiple spiders in the same process with custom settings could give some flexibility. It is explained on the same [page](https://docs.scrapy.org/en/latest/topics/practices.html). 
 


> The `crawl` function basically creates a list, which is the sent to the command line using the std lib `subprocess` module.

Yea I saw that... and TBH this subprocess module thing gave me tough time as I couldn't debug it further in vscode. 
May I ask, why you preffered using `subprocess` module instead of running it via scrapy's own crawl process:


Run Scrapy from a script[¶](https://docs.scrapy.org/en/latest/topics/practices.html#run-scrapy-from-a-script)
You can use the [API](https://docs.scrapy.org/en/latest/topics/api.html#topics-api) to run Scrapy from a script, instead of the typical way of running Scrapy via scrapy crawl.

Remember that Scrapy is built on top of the Twisted asynchronous networking library, so you need to run it inside the Twisted reactor.

The first utility you can use to run your spiders is [scrapy.crawler.CrawlerProcess](https://docs.scrapy.org/en/latest/topics/api.html#scrapy.crawler.CrawlerProcess). This class will start a Twisted reactor for you, configuring the logging and setting shutdown handlers. This class is the one used by all Scrapy commands.

Here’s an example showing how to run a single spider with it.

`import scrapy
from scrapy.crawler import CrawlerProcess

class MySpider(scrapy.Spider):
    # Your spider definition
    ...

process = CrawlerProcess(settings={
    "FEEDS": {
        "items.json": {"format": "json"},
    },
})

process.crawl(MySpider)
process.start() # the script will block here until the crawling is finished`

**Reply by eliasdabbas** [reply]
The `CrawlerProcess` was my first option actually. If I remember correctly, its problem was that it needed to be run in a new session. The idea behind this structure is to allow crawling to run within a Jupyter notebook/lab environment, and have users continue their work normally. Since it had to be restarted (all variables deleted), then I thought it would make sense to just go with the standard Scrapy implementation and run the same way it is run from the command line, with all options provided as is.

I might be wrong, and if there is a good reason to re-write it, and if there are issues with the current implementation we can review them and consider other options.

I'll try the proxy/meta implementation and will let you know.

**Reply by ska-ibees** [reply]
> I might be wrong, and if there is a good reason to re-write it, and if there are issues with the current implementation we can review them and consider other options.

One thing that's coming in mind is the lack of error reporting etc. (configuring the logging and setting shutdown handlers). I am unsure if we can catch any errors reported by the subprocesss. IMHO, it is something important within non-lab enviorment. Well, need further investigation to compare command line vs CrawlerProcess.



> I'll try the proxy/meta implementation and will let you know.

Alright. (BTW, I tested with a free account on  `webshare.io` )

**Reply by ska-ibees** [reply]
Well, i made several attempts to get errors/excemptions from the response but couldn't succeed.

Basically, I need to trigger some actions by having errors as call back.

One possibility would be to inherit/access:

```
[def errback(self, failure):
        if not failure.check(scrapy.exceptions.IgnoreRequest):
            self.logger.error(repr(failure))
            yield {'url': failure.request.url,
                   'crawl_time': datetime.datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S'),
                   'errors': repr(failure)}](https://github.com/eliasdabbas/advertools/blob/3ba5fe9a02965e1dc48cf23582217e3a5e111aaf/advertools/spider.py#L675)
```


Any idea about how to get it in return of adv.crawl() ?

**Reply by eliasdabbas** [reply]
> One thing that's coming in mind is the lack of error reporting etc. (configuring the logging and setting shutdown handlers)

There are several settings for logging, like the log format, log file and log date format for example. Is that what you meant?
There are also a few `CLOSESPIDER_*` settings. Unless you meant something else?

> Alright. (BTW, I tested with a free account on webshare.io )

I just tested it, and works fine. Great! 
I think another thing can be to allow users to introduce several proxies, and each request choses a random proxy. Since meta is flexible.
What do you think?

> Any idea about how to get it in return of adv.crawl() ?

You can trigger `errback` by using a wrong URL: 

`adv.crawl("doesnt_exit", "output_file.jl")`

It will create an errors column in the output file with the error details.

**Reply by ska-ibees** [reply]
> There are several settings for logging, like the log format, log file and log date format for example. Is that what you meant?
> There are also a few `CLOSESPIDER_*` settings. Unless you meant something else?

Alright, logging is clear now. Thanks



> I just tested it, and works fine. Great!
> I think another thing can be to allow users to introduce several proxies, and each request choses a random proxy. Since meta is flexible.
> What do you think?
IP reliability and type of proxy are critical factors here. So, I think let's keep it simple and generic and let proxy provider takes care of ip rotation. Like you see in case of webshare.io. 



> You can trigger `errback` by using a wrong URL:
> 
> `adv.crawl("doesnt_exit", "output_file.jl")`
> 
> It will create an errors column in the output file with the error details.

I see! So, the output_file.jl is  going to be a single source of truth... :) Thanks for pointing it.

**Comment by ska-ibees** [comment]
Finally some good news!!!

@eliasdabbas Please see this PR

https://github.com/eliasdabbas/advertools/pull/202/files

**Reply by eliasdabbas** [reply]
I like the more general approach!
We can use the `meta` parameter, which can include the proxy, and other stuff. 
Very nice. 

I'll check in more detail, and I'll get back to you. 
Thanks again.

**Reply by ska-ibees** [reply]
You are welcome. 

Yes, I intended to include native "meta" property from Core to take benefit out of it. Actually, it is equally powerful as of use of "custom-settings". Proxy is just one [meta attribute](https://github.com/eliasdabbas/advertools/pull/202/files) among others. 

> meta[¶](https://docs.scrapy.org/en/latest/topics/request-response.html#scrapy.http.Request.meta)
> A dict that contains arbitrary metadata for this request. This dict is empty for new Requests, and is usually populated by different Scrapy < components (extensions, middlewares, etc). So the data contained in this dict depends on the extensions you have enabled.
> 
> See [Request.meta special keys](https://docs.scrapy.org/en/latest/topics/request-response.html#topics-request-meta) for a list of special meta keys recognized by Scrapy.
> 
> This dict is [shallow copied](https://docs.python.org/3/library/copy.html) when the request is cloned using the copy() or replace() methods, and can also be accessed, in your spider, from the response.meta attribute.

**Reply by eliasdabbas** [reply]
@ska-ibees 
Just curious what you think about the multi proxy idea and randomization? 

Do you have any ideas? 

I think it could be something like this: 

```python
adv.crawl(url_list, output_file, follow_links=True,
         meta = {"proxy": ["http://127.0.0.1:3128", "http://127.0.0.2:3128", "http://127.0.0.3:3128", "http://127.0.0.4:3128"]}
```

for example. 

Each request, a random proxy will be selected from the available list.
It would automatically be included in the output file.

**Reply by ska-ibees** [reply]
@eliasdabbas  First of all forgive my ignorance that I didn't know much about you earlier. You don't know how much I feel honored to connect with you! Thanks for your contributions and contniued knowledge sharing.

**Reply by ska-ibees** [reply]
> Just curious what you think about the multi proxy idea and randomization?
> 
> Do you have any ideas?

AFAIK, IP rotation/multi IP proxy is not supported within Scrapy core. However,  it can be implemented using middleware. I have tested this one and it works fine[ Link ](https://github.com/TeamHG-Memex/scrapy-rotating-proxies)

Main problem that I see in this approach is that you have to manually update list of proxy IPs. Also, in order to avoid being banned IP's repute is important. That's why I earliler suggested to keep IP rotation outside of the scrapy. IMHO using a proxy API service is would be better for production use.

**Reply by eliasdabbas** [reply]
> @eliasdabbas First of all forgive my ignorance that I didn't know much about you earlier. You don't know how much I feel honored to connect with you! Thanks for your contributions and contniued knowledge sharing.

Thanks! Very kind of you. The honor is mine. 



> AFAIK, IP rotation/multi IP proxy is not supported within Scrapy core. However, it can be implemented using middleware. I have tested this one and it works fine[ Link ](https://github.com/TeamHG-Memex/scrapy-rotating-proxies)

Yes, this middleware is more extensive and flexible than what we are planning. Incorporating any Scrapy middleware can be a future project to work on also.

My suggestion is very simple actually. It uses the same approach you suggested, with reputable proxies. The user has the option of inputting a list of one or more proxies to the `crawl` function. 

Those will be sent, and used in making requests [here](https://github.com/eliasdabbas/advertools/blob/93055877b0b2cc14ad59acce6833cf127c462818/advertools/spider.py#L671) (and other places where relevant). This function call will instead of taking the given proxy will run `random.choice([proxy1, proxy2, proxy3, ...])`. 

If the user specified a single proxy, it will always be "chosen", otherwise a random proxy will be used for each request. 

Does that make sense?

**Reply by ska-ibees** [reply]
> Thanks! Very kind of you. The honor is mine.

To be honest, feeling blessed while working with you. 


> Does that make sense?


Got your point... so basically we are talking about a sort of wraper function within advertools instead of dealing it within Scrapy. 
```

For example: Re-constructing meta with a random proxy from the list. 

yield Request(url, callback=self.parse, errback=self.errback, meta=self.meta)
```
Yes, it does make lots of sense... Let me try and push PR for it.

**Reply by ska-ibees** [reply]
PR #207 to support selection of random proxy from the list of proxies.

**Comment by ska-ibees** [comment]
Syntax for testing multi proxy:


```
url_list = [
    "https://www.example.com", "https://www.example.com", "https://www.example.com"
]

meta = {
    "proxy": [
        "http://abc1.com:80",
        "http://abc2.com:80",
        "http://abc3.com:80",
        "http://abc4.com:80",
    ]
}
try:
            adv.crawl(
                url_list,
                output_file,
                follow_links=True,
                custom_settings={
                    "ROBOTSTXT_OBEY": True,
                    "USER_AGENT": "Mozilla/5.0 (iPad; CPU OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148",
                }, meta=meta,
            )
```
